external/concealer/include/concealer/autoware_user.hpp

```diff
@@ -0,0 +1,153 @@
+// Copyright 2015 TIER IV, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef CONCEALER__AUTOWARE_USER_HPP_
+#define CONCEALER__AUTOWARE_USER_HPP_
+
+// #define CONCEALER_ISOLATE_STANDARD_OUTPUT
+
+#include <sys/wait.h>
+
+#include <autoware_adapi_v1_msgs/msg/mrm_state.hpp>
+#include <autoware_auto_control_msgs/msg/ackermann_control_command.hpp>
+#include <autoware_auto_planning_msgs/msg/path_with_lane_id.hpp>
+#include <autoware_auto_system_msgs/msg/emergency_state.hpp>
+#include <autoware_auto_vehicle_msgs/msg/turn_indicators_command.hpp>
+#include <chrono>
+#include <concealer/launch.hpp>
+#include <concealer/task_queue.hpp>
+#include <concealer/transition_assertion.hpp>
+#include <concealer/utility/autoware_stream.hpp>
+#include <concealer/utility/visibility.hpp>
+#include <exception>
+#include <geometry_msgs/msg/accel.hpp>
+#include <geometry_msgs/msg/pose_stamped.hpp>
+#include <limits>
+#include <rclcpp/rclcpp.hpp>
+#include <traffic_simulator_msgs/msg/waypoints_array.hpp>
+#include <utility>
+
+namespace concealer
+{
+/* ---- NOTE -------------------------------------------------------------------
+ *
+ *  The magic class 'AutowareUser' is a class that makes it easy to work with
+ *  Autoware from C++. The main features of this class are as follows
+ *
+ *    (1) Launch Autoware in an independent process upon instantiation of the
+ *        class.
+ *    (2) Properly terminates the Autoware process started by the constructor
+ *        upon destruction of the class.
+ *    (3) Probably the simplest instructions to Autoware, consisting of
+ *        initialize, plan, and engage.
+ *
+ * -------------------------------------------------------------------------- */
+class AutowareUser : public rclcpp::Node
```

Comment:
I agree that this is not a very good naming, but I did not come up with something more universal then.

The motivation behind "User" naming is that this class fits the behavior of the person that:
- executes and closes the autoware launch file
- sends Autoware commands - init pose, plan, and engage
- verifies that Autoware is initializing step-by-step correctly
The implementation reflects that behavior.

This class additionally sends cooperate commands. I am unsure which code utilizes this, but since they are not needed on simple_sensor_simulator, I left it on this side of the concealer.

The design of this class is related to the things that will be needed on the simple_sensor_simulator side to simulate vehicle dynamics and vehicle state reports publishing. Of course, it might be changed, but in my opinion, the current design reflects this division, but of course, I might be missing something.

Therefore another naming might be in order. Possible alternatives that I now thought of are:
- `AutowareStartupCoordinator`
- `AutowareLaunchManager`
- `AutowareExecution`
Please let me know if one of those is more appropriate.

Reply:
I researched TIER IV products dealing with the topics and services that the current `AutowareUniverseUser` deals with.
The results of that research revealed that the lineup of topics and services matches well with a product called FOA (FOA; Field Operator Application) instead of FMS.

FOA is an application used by field operators to operate Autoware in experiments with real vehicles.
Since the scenario simulator is a tool used to simulate the demonstration experiment on a desk, it is highly semantically consistent that what the scenario simulator can tell the Autoware to do and what information it can obtain matches that of the field operator.

Therefore, I propose the following proposed renaming.
- Rename `AutowareUser` to `FieldOperatorApplication`
- Define `template <typename> class FieldOperatorApplicationFor;` with no implementation
- Rename `AutowareUniverseUser` to specialization `FieldOperatorApplicationFor<AutowareUniverse>`